### PR TITLE
Make `Foreach-Object` 2x faster by reducing unnecessary allocations and boxing

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Make `Foreach-Object` 2x faster by reducing unnecessary allocations and boxing.

1. Memory profiling of `$null = 1..3e5 | % { 1+1 }` shows huge allocation of ``System.Tuple`2`` and the boxed enum type `S.M.A.ConfigScope`. (See below)
They are resulted by the policy setting cache we chose for GPO: `ConcurrentDictionary<Tuple<ConfigScope, string>, PolicyBase>`. Every time we query for the GPO policy settings, an instance of `Tuple<ConfigScope, string>` has to be created. The created `Tuple` object is also short-living, only for the querying purpose. When the tuple is used as a key for querying, the enum value gets boxed for equality comparison and hash code calculation, because ``Tuple`2`` uses `EqualityComparer<object>.Default` as the comparer which treat items of the tuple as `object` (see [``Tuple`2.Equals()``](https://source.dot.net/#System.Private.CoreLib/shared/System/Tuple.cs,219) and [``Tuple`2.GetHashCode()``](https://source.dot.net/#System.Private.CoreLib/shared/System/Tuple.cs,259)).
To avoid those unnecessary allocations, the cache for GPO is changed to `ConcurrentDictionary<ConfigScope, ConcurrentDictionary<string, PolicyBase>>` to avoid using `Tuple` as the key.

2. Add code in `CompiledScriptBlock.cs` to skip calling `LogScriptBlockCreation` when logging is skipped.

3. Add a secondary cache for `ScriptBlockLogging` setting in `CompiledScriptBlock.cs` and a secondary cache for `Transcription` setting in `MshHostUserInterface.cs`, because they are used very frequently.

4. Memory profiling shows `28 mb` allocation of `System.Boolean`. This is caused by passing a bool value to a `object` type parameter when writing tracing message in `ParameterBinder`.

5. Memory profiling shows `28 mb` allocation of a closure class `<>c_DisplayClass55_0`. This is caused by [this piece of code](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/lang/scriptblock.cs#L978-L989). This is a hot code path and thus results in a lot allocation of the auto-generated closure. The `RunActionIfNoRunningPipelinesWithThreadCheck` method is refactored to avoid passing a delegate around.

#### Before the changes

image

```
PS> Measure-Command { 1..1e5 | % { 1+1 } > $null }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 508
Ticks             : 5085940
TotalDays         : 5.88650462962963E-06
TotalHours        : 0.000141276111111111
TotalMinutes      : 0.00847656666666667
TotalSeconds      : 0.508594
TotalMilliseconds : 508.594
```

#### After the changes

image

```
PS> Measure-Command { 1..1e5 | % { 1+1 } > $null }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 254
Ticks             : 2545267
TotalDays         : 2.94591087962963E-06
TotalHours        : 7.07018611111111E-05
TotalMinutes      : 0.00424211166666667
TotalSeconds      : 0.2545267
TotalMilliseconds : 254.5267
```

### BUT ...

```
PS> Measure-Command { 1..1e5 | & { process { 1+1 } } > $null }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 42
Ticks             : 426360
TotalDays         : 4.93472222222222E-07
TotalHours        : 1.18433333333333E-05
TotalMinutes      : 0.0007106
TotalSeconds      : 0.042636
TotalMilliseconds : 42.636
```

`Foreach-Object` is still 4 - 6 times slower than a filter function. This is mainly because:
1. `Foreach-Object` requires parameter binding for pipeline input while a filter function like the above one doesn't.
2. `Foreach-Object` calls into `ScriptBlock.InvokeUsingCmdlet` over and over again for every pipeline input. `InvokeUsingCmdlet` needs to do necessary setups/cleanup before/after invoking the script block and `Foreach-Object` has to pay that tax over and over again. While for a filter, the setup and cleanup is done only once because the invocation stays in the same function.
3. `Foreach-Object` runs the script block in the current scope so the compilation of the script block is un-optimized. While for a filter function, it runs in its own local scope and optimization is turned on.

### Next Step

Investigate to avoid unnecessary parameter binding for pipeline objects by using binders.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 1]
```diff
@@ -1919,17 +1943,16 @@ char ToLower(char c)
...
             {
-                forceLogCreation = true;
-            }
+                // When invoking, log the creation of the script block if it has suspicious content
+                bool forceLogCreation = scriptBlock._scriptBlockData.HasSuspiciousContent;
 
-            // We delay logging the creation util the 'Start' so that we can be sure we've
-            // properly analyzed the script block's security.
-            LogScriptBlockCreation(scriptBlock, forceLogCreation);
+                // We delay logging the creation util the 'Start' so that we can be sure we've
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: util?

File: stylecop.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
         },
         "namingRules" : {
             "allowCommonHungarianPrefixes" : true,
-            "allowedHungarianPrefixes" : [ "n", "r", "l", "i", "io", "is", "fs", "lp", "dw", "h", "rs", "ps", "op" ]
+            "allowedHungarianPrefixes" : [ "n", "r", "l", "i", "io", "is", "fs", "lp", "dw", "h", "rs", "ps", "op", "sb" ]
```
Comment (iSazonov) [inline stylecop.json]: Please add this in Settings.StyleCop too.
(Currently CodeFactor uses only Settings.StyleCop)

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Good catch. Will fix the typo.

File: stylecop.json (hunk 2 above)
Comment (daxian-dbw) [reply stylecop.json]: Will do.

**Review by TravisEz13 — APPROVED** [review]
with @iSazonov 's comments

File: src/System.Management.Automation/engine/CmdletParameterBinderController.cs [hunk 3]
```diff
@@ -3484,7 +3484,7 @@ private bool BindPipelineParametersPrivate(PSObject inputToOperateOn)
                         validParameterSets);
             }
 
-            s_tracer.WriteLine("aParameterWasBound = {0}", aParameterWasBound);
+            s_tracer.WriteLine("aParameterWasBound = {0}", aParameterWasBound.ToString());
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CmdletParameterBinderController.cs]: It's worth reviewing tracing messages like this to see if there are still useful. Without any context, this one doesn't look useful.

**Review by lzybkr — APPROVED** [review]

File: src/System.Management.Automation/engine/CmdletParameterBinderController.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/CmdletParameterBinderController.cs]: Will do the review, at least in the parameter binding code :)

Comment (iSazonov) [reply src/System.Management.Automation/engine/CmdletParameterBinderController.cs]: I suggest to add new overload like `internal void WriteLine(string format, bool arg1)`
To save time I created #10052. There is 18 references to the new overload and up to 46 references for other value types.
So we could revert the change in current PR.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 4]
```diff
@@ -1361,53 +1369,61 @@ internal bool Compile(bool optimized)
...
+                    scriptBlock.ScriptBlockData.IsProductCode)
                 {
-                    // If script block logging is explicitly disabled, or it's from a trusted
-                    // file or internal, skip logging.
-                    if (logSetting?.EnableScriptBlockLogging == false ||
-                        scriptBlock.ScriptBlockData.IsProductCode)
-                    {
-                        return;
-                    }
+                    scriptBlock.SkipLogging = true;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: It seems we skip logging (with line 1425) only if
```powershell
if (scriptBlock.ScriptBlockData.IsProductCode)
{
    scriptBlock.SkipLogging = true;
}
```
We could simplify the code and move this to line 1379.

File: src/System.Management.Automation/engine/CmdletParameterBinderController.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/CmdletParameterBinderController.cs]: OK, I will revert this line.

**Review by iSazonov — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**